src/libraries/System.Text.Json/src/System/Text/Json/Serialization/PooledByteBufferWriter.cs

```diff
@@ -138,7 +138,17 @@ private void CheckAndResizeBuffer(int sizeHint)
             {
                 int growBy = Math.Max(sizeHint, _rentedBuffer.Length);
 
-                int newSize = checked(_rentedBuffer.Length + growBy);
+                int newSize = _rentedBuffer.Length + growBy;
+
+                if ((uint)newSize > int.MaxValue && growBy > sizeHint)
```

Comment:
nit: Consider re-writing it as follows instead (smaller assembly and reduces the extra checks in the common case):
- Use a local for the length
- Nest the second check inside the first.
- Remove the `growBy > sizeHint` condition, and optimistically assume  that if `newSize` is too large its because we grew by `_rentedBuffer` and not by `sizeHint` and that growing by `sizeHint` might succeed.
```C#
int currentLength = _rentedBuffer.Length;
int growBy = Math.Max(sizeHint, currentLength);

int newSize = currentLength + growBy;

if ((uint)newSize > int.MaxValue)
{
    newSize = currentLength + sizeHint;
    if ((uint)newSize > int.MaxValue)
    {
        newSize = int.MaxValue;
    }
}
```

Reply:
Nice, I did the 2nd and 3rd suggestion some time ago and undid because was not sure if it was better.

Reply:
> Are all these new branches covered by tests? Can you check test coverage by running the following and seeing the coverage report generated (https://github.com/dotnet/runtime/blob/9900dfb4b2e32cf02ca846adaf11e93211629ede/docs/workflow/building/libraries/code-coverage.md#local-code-coverage-runs):
> 
> ```
> dotnet msbuild /t:BuildAndTest /p:Coverage=true
> ```

The tests seems to take forever to start when I enable Coverage, is it normal?

Reply:
> The tests seems to take forever to start and run when I enable Coverage, is it normal?

They should still start quickly, e.g. you should see the Discovering/Discovered/Starting lines quickly, but the tests can take a long time to run.  Here's what I get with what's in master:
```
    Discovering: System.Text.Json.Tests (method display = ClassAndMethod, method display options = None)
    Discovered:  System.Text.Json.Tests (found 1649 of 1680 test cases)
    Starting:    System.Text.Json.Tests (parallel test collections = on, max threads = 12)
    Finished:    System.Text.Json.Tests
  === TEST EXECUTION SUMMARY ===
     System.Text.Json.Tests  Total: 11344, Errors: 0, Failed: 0, Skipped: 0, Time: 58.468s
```